// grunt-npm2bower-sync/Gruntfile.js
/*
 * grunt-sync-pkg
 * https://github.com/jps/grunt-sync-pkg
 *
 * Copyright (c) 2013 Jon Schlinkert
 * Licensed under the MIT license.
 */

'use strict';

module.exports = function (grunt) {

  // load sync task
  grunt.loadTasks('tasks');

  // Project configuration.
  grunt.initConfig({
    jshint: {
      all: ['Gruntfile.js', 'tasks/*.js'],
      options: {
        "curly": true,
        "eqeqeq": true,
        "immed": true,
        "latedef": true,
        "newcap": true,
        "noarg": true,
        "sub": true,
        "undef": true,
        "boss": true,
        "eqnull": true,
        "node": true
      }
    },

    sync: {
      all: {
        options: {
          from: './package.json',
          to: './bower.json',
          sync: ['author', 'name', 'version', 'private', 'license']
        }
      }
    }
  });

  // These plugins provide necessary tasks.
  grunt.loadNpmTasks('grunt-contrib-jshint');

  // By default, lint and run all tests.
  grunt.registerTask('default', ['jshint', 'sync']);

};
